[AArch64] Re-enable rematerialization for streaming-mode-changing functions.

sdesmalen-arm opened the pull request:
We can add implicit defs/uses of the 'VG' register to the instructions to prevent the register allocator from rematerializing values in between streaming-mode changes, as the def/use of VG will further nail down the ordering that comes out of ISel. This avoids the heavy-handed approach to prevent any kind of rematerialization.

While we could add 'VG' as a Use to all SVE instructions, we only really need to do this for instructions that are rematerializable, as the smstart/smstop instructions and pseudos act as scheduling barriers which is sufficient to prevent other instructions from being scheduled in between the streaming-mode-changing call sequence. However, we may revisit this in the future.

sdesmalen-arm committed:
[AArch64] Re-enable rematerialization for streaming-mode-changing functions.

We can add implicit defs/uses of the 'VG' register to the instructions to
prevent the register allocator from rematerializing values in between
streaming-mode changes, as the def/use of VG will further nail down the
ordering that comes out of ISel. This avoids the heavy-handed approach to
prevent any kind of rematerialization.

While we could add 'VG' as a Use to all SVE instructions, we only really
need to do this for instructions that are rematerializable, as the
smstart/smstop instructions and pseudos act as scheduling barriers which is
sufficient to prevent other instructions from being scheduled in between
the streaming-mode-changing call sequence. However, we may revisit this in
the future.

paulwalker-arm commented:
The approach looks sensible to me.  I do think we should roll out the `VG` change across the other instructions, or at least those that directly access `VG` like `ADDVL`, but agree that's better done as follow on work.

sdesmalen-arm committed:
Remove undefined behaviour from lambda

paulwalker-arm commented on llvm/lib/Target/AArch64/AArch64InstrInfo.cpp:
```diff
@@ -9494,40 +9494,12 @@ bool AArch64InstrInfo::isReallyTriviallyReMaterializable(
-          }
-          return AArch64::ZPRRegClass.contains(MO.getReg()) ||
-                 AArch64::PPRRegClass.contains(MO.getReg());
+    if (any_of(MI.operands(), [&MFI](const MachineOperand &MO) {
+          return MO.isFI() &&
+                 MFI.getStackID(MO.getIndex()) == TargetStackID::ScalableVector;
```
Can you provide an example of a rematerialisable instruction that doesn't say it uses `VG` but has a scalable vector frame index.  I'm not saying there isn't one [1] but I'm wondering if there are specific cases that can be handled rather than having to walk every instructions operands.

[1] Perhaps `COPY FI:op`?

sdesmalen-arm replied:
The instruction that is rematerialisable that doesn't use VG (but would need it) is ADDXri, as this is used to materialise a frame-address, which only happens after register allocation. I've updated the PR to handle those instructions specifically, rather than walking through all instructions and their operands.

The corresponding test is `@call_to_non_streaming_pass_sve_objects` in llvm/test/CodeGen/AArch64/sme-streaming-interface.ll.

sdesmalen-arm committed:
Remove the need for isReallyTriviallyRematerializable...

...by moving the functionality to ADDXri, which is used for materializing
addresses to frame-objects, for example:

  gpr64:%dst = ADDXri fi:#0, 0

gets expanded to an actual addresses, by adding SP/FP + offset. If the
object is in the scalable vector region on the stack, then the offset will
be multiplied by 'mul VL', therefore reading VG.

This patch adds 'implicit $vg' to those instructions, which removes the
need for scanning all operands of all instructions for frame-indices.

I believe this is the only instruction where this is strictly required,
because other SVE instructions are either not rematerializable, or they
have an implicitly use of VG.

sdesmalen-arm commented on llvm/lib/Target/AArch64/AArch64RegisterInfo.cpp:
```diff
@@ -443,6 +443,9 @@ AArch64RegisterInfo::getStrictlyReservedRegs(const MachineFunction &MF) const {
       Reserved.set(SubReg);
   }
 
+  // VG cannot be allocated
+  Reserved.set(AArch64::VG);
```
I'm not sure it really matters. None of the tests fail so it seems valid to just mark it as reserved, even if it's not used when functions are not compiled for SVE or SME.
